usr/lib/common/key.c

```diff
@@ -7902,6 +7920,22 @@ CK_RV aes_set_default_attributes(TEMPLATE *tmpl, CK_ULONG mode)
     }
     value_attr = NULL;
 
+    /* If CKA_VALUE specified in base tmpl, add CKA_VALUE_LEN to tmpl */
+    if (template_attribute_find(basetmpl, CKA_VALUE, &value_attr)) {
+        keysize = value_attr->ulValueLen;
+        rc = build_attribute(CKA_VALUE_LEN, (CK_BYTE *)&keysize, sizeof(CK_ULONG), &len_attr);
```

Comment:
There are cases where CKA_VALUE_LEN is already specified by the user, i.e. for derive or Unwrap of an AES key. I guess in these cases CKA_VALUE is not specified, so your code would not do anything (as intended). But are there cases where you would overwrite a user supplied CKA_VALUE_LEN ?

Reply:
Hmm, this is indeed a good question ... I added a check to make sure basetmpl does not already contain CKA_VALUE_LEN, before adding it to tmpl (which is later merged into basetmpl). However, in every case I can think of, CKA_VALUE_LEN should in fact contain the correct length of CKA_VALUE.